src/utils/jwt-signature-verifier.ts

```diff
@@ -0,0 +1,235 @@
+   */
+  public fetchPublicKeys(): Promise<{ [key: string]: string }> {
+    const publicKeysExist = (typeof this.publicKeys !== 'undefined');
+    const publicKeysExpiredExists = (typeof this.publicKeysExpireAt !== 'undefined');
+    const publicKeysStillValid = (publicKeysExpiredExists && Date.now() < this.publicKeysExpireAt);
+    if (publicKeysExist && publicKeysStillValid) {
+      return Promise.resolve(this.publicKeys);
+    }
+
+    const client = new HttpClient();
+    const request: HttpRequestConfig = {
+      method: 'GET',
+      url: this.clientCertUrl,
+      httpAgent: this.app.options.httpAgent,
+    };
+    return client.send(request).then((resp) => {
+      if (!resp.isJson() || resp.data.error) {
+        // Treat all non-json messages and messages with an 'error' field as
+        // error responses.
+        throw new HttpError(resp);
+      }
+      if (Object.prototype.hasOwnProperty.call(resp.headers, 'cache-control')) {
+        const cacheControlHeader: string = resp.headers['cache-control'];
+        const parts = cacheControlHeader.split(',');
+        parts.forEach((part) => {
+          const subParts = part.trim().split('=');
+          if (subParts[0] === 'max-age') {
+            const maxAge: number = +subParts[1];
+            this.publicKeysExpireAt = Date.now() + (maxAge * 1000);
+          }
+        });
+      }
+      this.publicKeys = resp.data;
+      return resp.data;
+    }).catch((err) => {
+      if (err instanceof HttpError) {
+        let errorMessage = 'Error fetching public keys for Google certs: ';
+        const resp = err.response;
+        if (resp.isJson() && resp.data.error) {
+          errorMessage += `${resp.data.error}`;
+          if (resp.data.error_description) {
+            errorMessage += ' (' + resp.data.error_description + ')';
+          }
+        } else {
+          errorMessage += `${resp.text}`;
+        }
+        throw new SignatureVerifierError({
+          code: SignatureVerifierErrorCode.INTERNAL_ERROR,
+          message: errorMessage
+        });
+      }
+      throw err;
+    });
+  }
+}
+
+export class PublicKeySignatureVerifier implements SignatureVerifier {
+  private publicKeyProvider: PublicKeyFetcher;
+  private getKeyHandler: jwt.GetPublicKeyOrSecret;
+
+  constructor(clientCertUrl: string, private algorithm: jwt.Algorithm,
+    readonly app: FirebaseApp) {
+    if (!validator.isURL(clientCertUrl)) {
+      throw new Error(
+        'The provided public client certificate URL is an invalid URL.',
+      );
+    }
+    else if (!validator.isNonEmptyString(algorithm)) {
+      throw new Error('The provided JWT algorithm is an empty string.');
+    }
+
+    this.publicKeyProvider = new PublicKeyFetcher(clientCertUrl, app);
+    this.getKeyHandler = this.getKey.bind(this);
+  }
+
+  public verify(token: string): Promise<void> {
+    const error = new SignatureVerifierError({
```

Comment:
`isSignatureValid()` returns a boolean, if we throw this in `isSignatureValid` it will never false. Having it return a boolean will let us loop through multiple public keys if we ever needed to in the future. WDYT?

Reply:
See my latest comments. My recommendation is still to go with a `Promise<void>` return type (it will help reduce some of the duplication present now). To verify against multiple keys, we can  always do something like:

```
return Promise.any(keys.map((key) => verifySignature(token, key)));
```